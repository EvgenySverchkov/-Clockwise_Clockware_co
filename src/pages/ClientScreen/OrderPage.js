import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import PropTypes from "prop-types";

import {
  addCurrentOrderToState,
  addSuitableMasters,
  addTownsToState
} from "../../store/clientSide/data/actions";
import {changeOrderFormIsLoad} from "../../store/clientSide/services/actions";

import { SERVERDOMAIN } from "../../services/serverUrls";

import OrderForm from "../../forms/OrderForm";

function OrderFormClient({history}) {
  const state = useSelector(state=>{
    return {
      currentOrder: state.client_order_reduser.currentOrder,
      orderFormIsLoad: state.client_services.orderFormIsLoad,
      townsArr: state.client_order_reduser.townsArr,
    }
  });
  const dispatch = useDispatch();

  useEffect(function () {
    getTownsFromServerToState();
  }, []);
  function changeHandler(e) {
    let idx = e.target.name;
    dispatch(addCurrentOrderToState({
      ...state.currentOrder,
      [idx]: e.target.value,
    }));
  }
  function getTownsFromServerToState() {
    const headers = {
      Authorization: localStorage.getItem("token")
        ? "Bearer " + localStorage.getItem("token")
        : "",
      "Content-Type": "application/json",
    };
    fetch(`${SERVERDOMAIN}/towns`, { headers })
      .then((json) => json.json())
      .then((data) => dispatch(addTownsToState(data)));
  }
  function submitHandler(e) {
    e.preventDefault();
    let trgElem = e.target;
    if (
      !trgElem.town.value ||
      !trgElem.size.value ||
      !trgElem.name.value ||
      !trgElem.email.value ||
      !trgElem.time.value ||
      !trgElem.date.value
    ) {
      alert("Please, filling all gaps!!!");
      return false;
    }

    if(trgElem.name.value.length <= 3){
      alert("Name field must be at least 3 characters");
      return false;
    }

    let endOrderTime;
    let clientTimeHour = +state.currentOrder.time.match(/\d\d/);
    let clientTimeMin = state.currentOrder.time.match(/:\d\d$/);
    switch (state.currentOrder.size) {
      case "small":
        endOrderTime = clientTimeHour + 1 + clientTimeMin;
        break;
      case "middle":
        endOrderTime = clientTimeHour + 2 + clientTimeMin;
        break;
      case "large":
        endOrderTime = clientTimeHour + 3 + clientTimeMin;
        break;
      default:
        endOrderTime = 0;
    }
    dispatch(changeOrderFormIsLoad(true));
    getFreeMastersByClientTownFromServer(
      SERVERDOMAIN,
      trgElem.town.value,
      state.currentOrder.time,
      endOrderTime,
      state.currentOrder.date
    ).then((data) => {
      dispatch(changeOrderFormIsLoad(false));
      if (data.success) {
        dispatch(addSuitableMasters(data.payload));
        history.push("/client/masters");
      } else {
        dispatch(addSuitableMasters([]));
        alert(data.msg);
        history.push("/client");
      }
    });
  }
  function getFreeMastersByClientTownFromServer(
    url,
    clientTown,
    clientTimeStart,
    clientTimeEnd,
    clientDate
  ) {
    let obj = {
      town: clientTown,
      timeStart: clientTimeStart,
      timeEnd: clientTimeEnd,
      date: clientDate,
    };
    return fetch(`${url}/freeMasters`, {
      method: "POST",
      headers: {
        Authorization: localStorage.getItem("token")
          ? "Bearer " + localStorage.getItem("token")
          : "",
        "Content-Type": "application/json;charset=utf-8",
      },
      body: JSON.stringify(obj),
    }).then((json) => json.json());
  }
  return (
    <>
      <div className="text-center mb-1">
        You are welcomed by <h1 className="font-italic">Clockwise company</h1>
      </div>
      <br />
      <div className="text-center">Fill out this form to order a master</div>
      <OrderForm
        submitHandler={submitHandler}
        changeHandler={changeHandler}
        currentOrder={state.currentOrder}
        isLoadOrderForm={state.orderFormIsLoad}
        townsArr={state.townsArr}
      />
    </>
  );
}

OrderFormClient.propTypes = {
  history: PropTypes.object,
};

export default OrderFormClient;
