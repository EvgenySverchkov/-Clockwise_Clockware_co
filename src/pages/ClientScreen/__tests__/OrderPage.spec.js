import React from "react";
import {create, act} from "react-test-renderer";
import {Provider} from "react-redux";
import configMockStore from "redux-mock-store";

import OrderPage from "../OrderPage";
import Context from "../../../ContextComponent";

const mockSrore = configMockStore();

const constantDate = new Date('2017-06-13')

Date = class extends Date {
  constructor() {
    return constantDate
  }
}

it("Test of <OrderPage/>", async ()=>{
    
    global.fetch = jest.fn(()=>Promise.resolve({json: jest.fn(()=>Promise.resolve())}));
    const store = mockSrore({
        ordersReducer: {
            currentOrder: {},
            orderFormIsLoad: false
        },
        townReduser: {
            townsArr: [{id: 1, name: "Town1"}, {id: 2, name: "Town2"}],
            townsInOrderFormIsLoad: false
        }
    })
    const mockContextValue = {
        openErrorWindowWithMsg: jest.fn(),
        openWarningTooltip: jest.fn(),
        warningTooltipMsg: "Warning"
    }
    const mockProps = {
        history: {
            push: jest.fn()
        }
    }
    let component;
    await act(async ()=>{
        component = await create(
        <Provider store = {store}>
            <Context.Provider value = {mockContextValue}>
                <OrderPage {...mockProps}/>
            </Context.Provider>
        </Provider>
        );
    });

    expect(component.toJSON()).toMatchSnapshot();

});