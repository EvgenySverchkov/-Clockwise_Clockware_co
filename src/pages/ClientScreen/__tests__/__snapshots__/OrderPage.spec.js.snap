// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of <OrderPage/> 1`] = `
Array [
  <div
    className="text-center mb-1"
  >
    You are welcomed by 
    <h1
      className="font-italic"
    >
      Clockwise company
    </h1>
  </div>,
  <br />,
  <div
    className="text-center"
  >
    Fill out this form to order a master
  </div>,
  <form
    className="mt-4 row justify-content-center"
    onBlur={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="form-group row text-center text-sm-left col-12 col-md-10 col-lg-8"
    >
      <label
        className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label font-weight-bold"
        htmlFor="name"
      >
        Enter 
        name
      </label>
      <div
        className="col-sm-8 col-md-8"
      >
        <input
          aria-describedby={null}
          className="form-control"
          id="name"
          name="name"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          required={true}
          title="Warning"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="form-group row text-center text-sm-left col-12 col-md-10 col-lg-8"
    >
      <label
        className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label font-weight-bold"
        htmlFor="email"
      >
        Enter 
        email
      </label>
      <div
        className="col-sm-8 col-md-8"
      >
        <input
          aria-describedby={null}
          className="form-control"
          id="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          required={true}
          title="Warning"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="form-group  text-center text-sm-left col-12 col-md-10 col-lg-8"
    >
      <div
        className="mb-2 font-weight-bold"
      >
        Choose size of clock
      </div>
      <div
        className="form-check form-check-inline"
      >
        <input
          className="form-check-input"
          id="smallSize"
          name="size"
          onChange={[Function]}
          type="radio"
          value="small"
        />
        <label
          className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label "
          htmlFor="smallSize"
        >
          Small
        </label>
      </div>
      <div
        className="form-check form-check-inline"
      >
        <input
          className="form-check-input"
          id="middleSize"
          name="size"
          onChange={[Function]}
          type="radio"
          value="middle"
        />
        <label
          className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label "
          htmlFor="middleSize"
        >
          Middle
        </label>
      </div>
      <div
        className="form-check form-check-inline"
      >
        <input
          className="form-check-input"
          id="largeSize"
          name="size"
          onChange={[Function]}
          type="radio"
          value="large"
        />
        <label
          className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label "
          htmlFor="largeSize"
        >
          Large
        </label>
      </div>
    </div>
    <div
      className="form-group  text-center text-sm-left col-12 col-md-10 col-lg-8"
    >
      <div
        className="mb-2 font-weight-bold"
      >
        Choose town
      </div>
      <div
        className="form-check-inline"
      >
        <div
          className="form-check form-check-inline"
        >
          <input
            className="form-check-input"
            id="Town1"
            name="town"
            onChange={[Function]}
            type="radio"
            value="Town1"
          />
          <label
            className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label "
            htmlFor="Town1"
          >
            Town1
          </label>
        </div>
      </div>
      <div
        className="form-check-inline"
      >
        <div
          className="form-check form-check-inline"
        >
          <input
            className="form-check-input"
            id="Town2"
            name="town"
            onChange={[Function]}
            type="radio"
            value="Town2"
          />
          <label
            className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label "
            htmlFor="Town2"
          >
            Town2
          </label>
        </div>
      </div>
    </div>
    <div
      className="form-group  text-center text-sm-left col-12 col-md-10 col-lg-8"
    >
      <div
        className="mb-2 font-weight-bold"
      >
        Choose date and time
        <br />
        <sub>
          *time from 8 to 18
        </sub>
      </div>
      <input
        aria-describedby={null}
        className="mr-1"
        id="date"
        max="2018-06-13"
        min="2017-06-12"
        name="date"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        required={true}
        title="Warning"
        type="date"
        value=""
      />
      <input
        aria-describedby={null}
        className=""
        id="time"
        max={null}
        min={null}
        name="time"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        required={true}
        title="Warning"
        type="time"
        value=""
      />
    </div>
    <div
      className="row justify-content-sm-center col-12"
    >
      <input
        className="btn btn-primary col-12 col-sm-4 mt-3"
        type="submit"
        value="Next step"
      />
    </div>
  </form>,
]
`;
